packages/cli/src/util/progress.ts

```diff
@@ -0,0 +1,55 @@
+// zero is considered first index in the range
+type NeedleFunc = (needle: number) => void;
+type ProgressFunc = (opts: {current: number; total: number; percentage: number; ratePerSec: number}) => void;
+
+export function showProgress({
+  total,
+  signal,
+  frequencyMs,
+  progress,
+}: {
+  total: number;
+  signal: AbortSignal;
+  frequencyMs: number;
+  progress: ProgressFunc;
+}): NeedleFunc {
+  let current = 0;
+  let last = 0;
+  let lastProcessTime: number = Date.now();
+
+  const needle: NeedleFunc = (needle: number) => {
+    // zero is considered first index in the range
+    current = needle + 1;
+
+    if (current >= total) {
+      processProgress();
+    }
+  };
+
+  const processProgress = (): void => {
+    const currentTime = Date.now();
+    const processTime = currentTime - lastProcessTime;
+
+    progress({
+      current,
+      total,
+      ratePerSec: processTime === 0 ? 0 : ((current - last) / processTime) * 1000,
```

Comment:
unless we need to calculate the rate per second, we should use rate per minute directly and make sure to use floating point numbers. currently the rate per minute is always a multiple of 6:

- 0.00
- 6.00
- 12.00
- 18.00